testindex/nuts_and_bolts.py

```diff
@@ -0,0 +1,157 @@
+from datetime import datetime, date
+from glob import glob
+from os import environ, path, mkdir, unsetenv, listdir, unlink, devnull, getenv
+from shutil import rmtree
+from subprocess import check_call, CalledProcessError, STDOUT
+
+from yaml import load
+
+from dependency_validation import DependencyValidator
+from summary import Summary
+
+
+def execute_command(cmd):
+    """Execute a specified command"""
+    try:
+        fnull = open(devnull, "w")
+        check_call(cmd, stdout=fnull, stderr=STDOUT)
+        return True
+    except CalledProcessError:
+        return False
+
+
+class Environment(object):
+    """Handles the bringing up and teardown of the test environment"""
+
+    def __init__(self):
+
+        home = getenv("HOME")
+        if home:
+            dump_location = home + "/cccp-index-test"
+        else:
+            dump_location = "./cccp-index-test"
+
+        if not path.isabs(dump_location):
+            dump_location = path.abspath(dump_location)
+
+        self.dump_directory = dump_location
+        self.index_test_bench = dump_location + "/index_test_bench"
+        time_stamp = datetime.now().strftime("%Y-%m-%d_%H:%M:%S:%s")
+        self.index_dir = self.index_test_bench + "/" + time_stamp
+        self.test_index = self.index_dir + "/index.d"
+        self.summary_location = self.index_dir + "/summary.log"
+        self.repo_dump = dump_location + "/repositories"
+        generator_location = dump_location + "_generator_ref/"
+        self.generator_dir = generator_location + time_stamp
+        self.silent = False
+
+        self.old_environ = dict(environ)
+
+        if not path.exists(self.dump_directory):
+            mkdir(self.dump_directory)
+
+        if not path.exists(self.index_test_bench):
+            mkdir(self.index_test_bench)
+
+        if not path.exists(self.index_dir):
+            mkdir(self.index_dir)
+
+        if not path.exists(self.test_index):
+            mkdir(self.test_index)
+
+        if not path.exists(self.repo_dump):
+            mkdir(self.repo_dump)
+
+        if not path.exists(generator_location):
+            mkdir(generator_location)
+
+        if not path.exists(self.generator_dir):
+            mkdir(self.generator_dir)
+
+        if path.exists(self.generator_dir):
+            self._cleanup_content(self.generator_dir)
+
+        if not path.exists(self.generator_dir):
+            mkdir(self.generator_dir)
+
```

Comment:
Ok i see your point though i dont need 2 checks for generator dir anymore since i am now time stamping it and hence each run gets its own generator dir. So i am surrounding it with a for loop and try catch.